### PR TITLE
[FIRRTL] Rewrite ExtractClasses as LowerClasses.

### PR DESCRIPTION
**Pull request by mikeurbach**
This is a fairly significant rewrite, so I've chosen to rename the pass, but otherwise share some of the good parts. A lot of warts are hopefully removed.

The pass lowers both FIRRTL modules with property ports as well as FIRRTL classes into OM classes. It does a linear walk to initially create the OM classes, and then processes each in parallel.

The logic to move ops into the new class body is greatly simplified--ops are copied over, in order, if they have property operands or results. We might make this even simpler with a trait at some point.

Instantiations and the instance graph no longer need to be kept in sync--only FIRRTL Object instantiations are affected. These are converted to OM Object instantiations. It's actually impossible to represent a FIRRTL Object that takes inputs with the current flow-checking. I tried to write out an implementation for it, even though I can't test it.

Besides the structural changes, I've also added type conversion from the FIRRTL ops to the OM versions. I'm using DialectConversion for this, which actually helps for this sort of thing. There are fairly simple patterns for converting the constant property ops to the OM versions. These could almost be templated on the operation class, but minor differences made that more trouble than its worth. There are also patterns for legalizing OM class fields and OM classes, which are just there to let the type converter do its thing.

I've added a basic unit test for the kind of IR we want to support here. Once more parser and flow-checking support is in, I will also include an end-to-end test from FIRRTL through ExportVerilog to the final MLIR.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+      formalParamNames.push_back(port.name);
+
+  OpBuilder builder = OpBuilder::atBlockEnd(getOperation().getBody(0));
+
+  // Take the name from the FIRRTL Class or Module to create the OM Class name.
+  StringRef className = moduleLike.getName();
+
+  // If the op is a Module, the OM Class would conflict with the HW Module, so
+  // give it a suffix. There is no formal ABI for this yet.
+  StringRef suffix = isa<FModuleOp>(moduleLike) ? "_Class" : "";
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:135]: This probably needs to unique the new name in a namespace...

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+           op->getRegion(0).getOps<firrtl::ObjectOp>())) {
+    // Collect its input actual parameters by finding any subfield ops that are
+    // assigned to. Take the source of the assignment as the actual parameter.
+    SmallVector<Value> actualParameters;
+    for (auto *user : firrtlObject->getUsers())
+      if (auto subfield = dyn_cast<ObjectSubfieldOp>(user))
+        for (auto *subfieldUser : subfield->getUsers())
+          if (auto propassign = dyn_cast<PropAssignOp>(subfieldUser))
+            if (propassign.getDest() == subfield.getResult())
+              actualParameters.push_back(propassign.getSrc());
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:246]: I couldn't actually write a test for this... but I think this is how it should work? Maybe I should remove it for now.

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/OM/OMOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Threading.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/raw_ostream.h"
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/Support/LogicalResult.h>
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:25]: nit: 
```suggestion
#include "mlir/IR/PatternMatch.h"
#include "mlir/Support/LogicalResult.h"
```

**Review by seldridge — COMMENTED** [review]
A few comments after a very quick read.

I like the use of the `DialectConversion` here. I had some questions about the motivation for combining the `firrtl.module` -> `[firrtl.module, firrtl.class]` with the conversion of `firrtl.class -> om.class`. However, that's both: (1) already reasonably factored in the existing pass and (2) not a big change from how the pass already works. I am not clear on the benefits of doing partial lowering in passes vs. full conversion in a monolithic pass (`LowerToHW`).

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+
+/// Lower FIRRTL Class and Object ops to OM Class and Object ops
+void LowerClassesPass::runOnOperation() {
+  MLIRContext *ctx = &getContext();
+
+  // Get the CircuitOp.
+  auto circuits = getOperation().getOps<CircuitOp>();
+  if (circuits.empty())
+    return;
+  CircuitOp circuit = *circuits.begin();
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:76]: om::ClassOp is currently defined to live within builtin::ModuleOp. we could change that, but this is why it's a ModuleOp pass right now. we have this assumption in a lot of places, so I'm fine with just erroring if there isn't one CircuitOp.

**Comment by mikeurbach** [comment]
> I had some questions about the motivation for combining the firrtl.module -> [firrtl.module, firrtl.class] with the conversion of firrtl.class -> om.class

Good question, I wrote this down somewhere else. The issue is once we convert all FIRRTL Classes to OM Classes, we can’t have any FIRRTL Object instantiations--they need to be converted to OM Object instantiations. And if any FIRRTL module instantiates an OM Class (which we want in the top-level FIRRTL module), it can’t pass the instantiated OM Object out its ports (which we also want in the top-level FIRRTL module). FIRRTL ports need to be property type, but the set of property types is sealed in the FIRRTL dialect. So we either need to allow OM Objects in FIRRTL ports as a foreign type, or do the conversion of FIRRTL modules while we convert FIRRTL classes.

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 5]
```diff
@@ -0,0 +1,413 @@
+//===- LowerClasses.cpp - Extract OM classes ----------------------------===//
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:1]: LowerClasses

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+using namespace circt::firrtl;
+using namespace circt::om;
+
+/// Helper class to capture details about a property.
+struct Property {
+  size_t index;
+  StringRef name;
+  Type type;
+  Location loc;
+};
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:38]: might want to put these in an anonymous namespace anyways

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+  // Get the CircuitOp.
+  auto circuits = getOperation().getOps<CircuitOp>();
+  if (std::distance(circuits.begin(), circuits.end()) != 1) {
+    getOperation().emitError("expected exactly one CircuitOp, but found ")
+        << std::distance(circuits.begin(), circuits.end());
+    return signalPassFailure();
+  }
+  CircuitOp circuit = *circuits.begin();
+
+  // Create new OM Class ops seriallly.
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:81]: `s@llly@lly`

**Comment by mikeurbach** [comment]
I'm going to go ahead an merge this so we have a baseline checked in. I can address any post merge review tomorrow.

File: lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp (hunk 2 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:251]: I think this assumes that the ports are assigned in-order. You'll need to be able to map from old-port-index to new-argument-index, and then check what port was being assigned by looking at the subfield op's index.

Comment (mikeurbach) [reply lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp:251]: Yeah makes sense, I wasn't actually able to test this part anyway, so it's sort of a placeholder.

**Pull request merged**